File: src/main/java/com/somefrills/features/tweaks/NoGhostBlocks.java
package com.somefrills.features.tweaks;

import com.somefrills.config.FrillsConfig;
import com.somefrills.config.tweaks.TweaksCategory.NoGhostBlocksConfig;
import com.somefrills.events.BreakBlockEvent;
import com.somefrills.events.PlaceBlockEvent;
import com.somefrills.features.core.Feature;
import com.somefrills.features.core.FrillsFeature;
import meteordevelopment.orbit.EventHandler;
import net.minecraft.block.BlockState;

import static com.somefrills.Main.mc;

@FrillsFeature
public class NoGhostBlocks extends Feature {
    private final NoGhostBlocksConfig config;

    public NoGhostBlocks() {
        super(FrillsConfig.tweaks.noGhostBlocks.enabled);
        config = FrillsConfig.tweaks.noGhostBlocks;
    }

    @EventHandler
    private void onBreakBlock(BreakBlockEvent event) {
        if (mc.isInSingleplayer() || mc.world == null || !config.breaking) return;
        event.cancel();
        BlockState blockState = mc.world.getBlockState(event.blockPos);
        blockState.getBlock().onBreak(mc.world, event.blockPos, blockState, mc.player);
    }

    @EventHandler
    private void onPlaceBlock(PlaceBlockEvent event) {
        if (!config.placing) return;
        event.cancel();
    }
}

File: src/main/kotlin/com/somefrills/features/tweaks/NoGhostBlocks.kt
package com.somefrills.features.tweaks

import com.somefrills.Main.mc
import com.somefrills.config.FrillsMod

import com.somefrills.config.tweaks.TweaksCategory.NoGhostBlocksConfig
import com.somefrills.events.BreakBlockEvent
import com.somefrills.events.PlaceBlockEvent
import com.somefrills.features.core.Feature
import com.somefrills.modules.FrillsFeature
import com.somefrills.events.core.EventHandle
import net.minecraft.world.level.block.state.BlockState

@FrillsFeature
object NoGhostBlocks : Feature(FrillsMod.config.tweaks.noGhostBlocks.enabled) {
    @JvmStatic
    val config: NoGhostBlocksConfig
        get() = FrillsMod.config.tweaks.noGhostBlocks

    @EventHandle
    private fun onBreakBlock(event: BreakBlockEvent) {
        val world = mc.level ?: return
        val player = mc.player ?: return
        if (mc.isSingleplayer || !config.breaking) return
        event.cancel()
        val blockState: BlockState = world.getBlockState(event.blockPos)
        blockState.block.playerWillDestroy(world, event.blockPos, blockState, player)
    }

    @EventHandle
    private fun onPlaceBlock(event: PlaceBlockEvent) {
        if (!config.placing) return
        event.cancel()
    }
}
